plugins/custom-dashboard/__init__.py

```diff
@@ -0,0 +1,913 @@
+"""
+Custom Dashboards.
+
+| Copyright 2017-2024, Voxel51, Inc.
+| `voxel51.com <https://voxel51.com/>`_
+|
+"""
+
+import enum
+import numpy as np
+import random
+from textwrap import dedent
+
+import fiftyone as fo
+import fiftyone.operators as foo
+import fiftyone.operators.types as types
+import fiftyone.types as fot
+import fiftyone.core.fields as fof
+from fiftyone import ViewField as F
+
+
+class PlotlyPlotType(enum.Enum):
+    BAR = 'bar'
+    SCATTER = 'scatter'
+    LINE = 'line'
+    PIE = 'pie'
+
+
+class PlotType(enum.Enum):
+    CATEGORICAL_HISTOGRAM = 'categorical_histogram'
+    NUMERIC_HISTOGRAM = 'numeric_histogram'
+    LINE = 'line'
+    SCATTER = 'scatter'
+    PIE = 'pie'
+
+
+NUMERIC_TYPES = (fof.IntField, fof.FloatField)
+
+CATEGORICAL_TYPES = (fof.StringField, fof.BooleanField)
+
+
+def get_plotly_plot_type(plot_type):
+    return {
+        PlotType.CATEGORICAL_HISTOGRAM: PlotlyPlotType.BAR,
+        PlotType.NUMERIC_HISTOGRAM: PlotlyPlotType.BAR,
+        PlotType.LINE: PlotlyPlotType.LINE,
+        PlotType.SCATTER: PlotlyPlotType.SCATTER,
+        PlotType.PIE: PlotlyPlotType.PIE,
+    }.get(plot_type)
+
+
+def get_plotly_config_and_layout(plot_config):
+    layout = {}
+    if plot_config.get('plot_title'):
+        # rely on the dashboard title for now
+        # layout['title'] = plot_config.get('plot_title')
+        pass
+    if plot_config.get('color'):
+        layout['marker'] = {'color': plot_config['color'].get('hex')}
+    if plot_config.get('xaxis'):
+        layout['xaxis'] = plot_config.get('xaxis')
+    if plot_config.get('yaxis'):
+        layout['yaxis'] = plot_config.get('yaxis')
+    if plot_config.get('plot_type') == 'numeric_histogram':
+        layout['bargap'] = 0
+        layout['bargroupgap'] = 0
+
+    return {'config': {}, 'layout': layout}
+
+
+REQUIRES_X = [PlotType.SCATTER, PlotType.LINE, PlotType.NUMERIC_HISTOGRAM]
+REQUIRES_Y = [PlotType.SCATTER, PlotType.LINE]
+
+
+def get_root(dataset, path):
+    root = None
+    schema = dataset.get_field_schema(flat=True)
+    for _path, field in schema.items():
+        if path.startswith(_path + ".") and isinstance(field, fo.ListField):
+            root = _path
+
+    return root
+
+
+def get_fields_with_type(dataset, field_types, root=None):
+    schema = dataset.get_field_schema(flat=True)
+    paths = []
+    for path, field in schema.items():
+        if root is not None and not path.startswith(root + "."):
+            continue
+
+        if isinstance(field, field_types) or (
+            isinstance(field, fo.ListField)
+            and isinstance(field.field, field_types)
+        ):
+            paths.append(path)
+
+    return paths
+
+def get_view_for_category(field, category, view):
```

Comment:
Okay this is a bit of a mouthful, but here's how to make this work for all possible types of fields the chart may be plotting:

```py
def get_view_for_value(sample_collection, path, value):
    """Returns a view into the given `sample_collection` that matches the given
    `value` within the given `path`.

    Supports label fields, list fields, and a combination of both.
    """
    root, leaf = _parse_path(sample_collection, path)
    is_label_field = _is_field_type(sample_collection, root, fo.Label)
    is_list_field = _is_field_type(sample_collection, path, fo.ListField)

    if is_label_field:
        if is_list_field:
            expr = F(leaf).exists() & F(leaf).contains(value)
        else:
            expr = F(leaf) == value

        return sample_collection.filter_labels(root, expr)

    if is_list_field:
        expr = F(path).exists() & F(path).contains(value)
    else:
        expr = F(path) == value

    return sample_collection.match(expr)


def _is_field_type(sample_collection, path, field_or_type):
    field = sample_collection.get_field(path)

    if issubclass(field_or_type, fo.Field):
        return isinstance(field, field_or_type)

    return (
        isinstance(field, fo.EmbeddedDocumentField)
        and issubclass(field.document_type, field_or_type)
    )


def _parse_path(sample_collection, path):
    if "." not in path:
        return path, None

    chunks = path.split(".")
    root = chunks[0]
    leaf = chunks[-1]

    # Handle dynamic documents
    idx = 0
    while (
        _is_field_type(sample_collection, root, fo.EmbeddedDocumentField)
        and not _is_field_type(sample_collection, root, fo.Label)
    ):
        idx += 1
        root += "." + chunks[idx]

    return root, leaf
```

Example usage:

```py
import fiftyone as fo
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart")

dataset[:100].tag_samples("hi")
dataset[:100].tag_labels("hi", label_fields="ground_truth")

view1 = get_view_for_value(dataset, "filepath", dataset.first().filepath)
view2 = get_view_for_value(dataset, "tags", "hi")
view3 = get_view_for_value(dataset, "ground_truth.detections.label", "cat")
view4 = get_view_for_value(dataset, "ground_truth.detections.tags", "hi")
```

Reply:
Clean.